046-analyse-broker-metrics-before-rolling-kafka.md

```diff
@@ -0,0 +1,112 @@
+This proposal describes a solution for KafkaRoller's shortcoming with brokers in log recovery.
+
+## Current situation
+
+The current logic in KafkaRoller:
+- Take a list of Kafka pods and reorder them based on the readiness.
+- Roll unready pods first without considering why it could be unready.
+- If the pod is the controller, roll it last.
+- Do not roll the pod, if it has an impact on the availability such as causing under replicated partitions.
+- If a Kafka pod does not become ready within the operational timeout (300000ms by default) after restarting, then it may force restart it without considering why it's taking a long time.
+
+The current flow of KafkaRoller
+![The current flow of KafkaRoller](images/046-kafka-roller-current-flow.png)
+
+A Kafka broker can take a long time to become ready while performing log recovery. In this case, KafkaRoller could end up force restarting the broker and this can continue indefinitely, because of the log recovery on the broker startup again. KafkaRoller currently does not have a way to know what state a broker is in before making a decision to force restart it.
+
+KafkaAgent currently collects the `kafka.server:type=KafkaServer,name=BrokerState` metric and creates a `kafka-ready` file on disk if the broker state value is greater than `3`(`RUNNING`) and not `127` (`UNKNOWN`). The Kafka readiness check passes if it finds the file on disk.
+
+## Motivation
+
+To address the issues with restarts and log recovery, KafkaRoller needs to be aware of the current broker state.  If a broker is performing log recovery, KafkaRoller should fail the reconciliation instead of force restarting it.
+
+## Proposal
+ 
+The objective of this proposal is to introduce a REST API endpoint within KafkaAgent, which exposes metrics from a Kafka broker. Furthermore, KafkaRoller would be modified so that it can query this endpoint to determine the current state of a broker, thus preventing unnecessary restarts. Additionally, KafkaRoller would allow a human operator to observe whether the log recovery process is making progress or is stuck.
+
+The following implementation details will describe behaviour of the new API in KafkaAgent and how KafkaRoller would interact with it and handle a broker that is in log recovery.
+
+### Implementation details
+
+**KafkaAgent**:
+
+- Add a web server to accept SSL connections on port 8443.
+- Use Jetty for the web server since Kafka already uses it therefore avoids an additional dependency.
+- Reuse the broker's own certificate that would be in the container for TLS authentication.
+- Configure the server with mTLS authentication.
+- Expose an endpoint `v1/broker-state` that includes the API version in the URI path.
+- Extend KafkaAgent to collect `kafka.log:type=LogManager,name=remainingLogsToRecover` and `kafka.log:type=LogManager,name=remainingSegmentsToRecover` metrics.
+- Use Jetty's native ServerHandler API to handle a `GET` request and return a JSON response that would look like this:
```

Comment:
just for completeness with the next two returns, let's add "return HTTP 200 (OK) with a body JSON response ..."